c-dependencies/js-compute-runtime/js-compute-builtins.cpp

```diff
@@ -1210,6 +1212,1279 @@ namespace NativeStreamSource {
+    RootedObject readableController(cx, JS::ReadableStreamGetController(cx, readable));
+    MOZ_ASSERT(readableController);
+
+    // 3.  If ! [ReadableStreamDefaultControllerCanCloseOrEnqueue](readableController) is false,
+    // throw a `TypeError` exception.
+    if (!JS::CheckReadableStreamControllerCanCloseOrEnqueue(cx, readableController, "enqueue")) {
+      return false;
+    }
+
+    // 4.  Let enqueueResult be ReadableStreamDefaultControllerEnqueue(readableController, chunk).
+    bool enqueueResult = JS::ReadableStreamEnqueue(cx, readable, chunk);
+
+    // 5.  If enqueueResult is an abrupt completion,
+    if (!enqueueResult) {
+      // 5.1.  Perform
+      // TransformStreamErrorWritableAndUnblockWrite(stream, enqueueResult.[Value]).
+      RootedValue resultValue(cx);
+      if (!JS_GetPendingException(cx, &resultValue)) {
+        return false;
+      }
+      JS_ClearPendingException(cx);
+
+      if (!TransformStream::ErrorWritableAndUnblockWrite(cx, stream, resultValue)) {
+        return false;
+      }
+
+      //     2.  Throw stream.[readable].[storedError].
+      RootedValue storedError(cx, JS::ReadableStreamGetStoredError(cx, readable));
+      JS_SetPendingException(cx, storedError);
+      return false;
+    }
+
+    // 6.  Let backpressure be ReadableStreamDefaultControllerHasBackpressure(readableController).
+    // (Inlined)
+    bool backpressure = !JS::ReadableStreamControllerShouldCallPull(cx, readableController);
+
+    // 7.  If backpressure is not stream.[backpressure],
+    if (backpressure != TransformStream::backpressure(stream)) {
+      //     1.  Assert: backpressure is true.
+      MOZ_ASSERT(backpressure);
+
+      //     2.  Perform ! TransformStreamSetBackpressure(stream, true).
+      TransformStream::SetBackpressure(cx, stream, true);
+    }
+
+    return true;
+  }
+
+  /**
+   * TransformStreamDefaultControllerTerminate
+   */
+  bool Terminate(JSContext* cx, HandleObject controller) {
+    MOZ_ASSERT(is_instance(controller));
+
+    // 1.  Let stream be controller.[stream].
+    RootedObject stream(cx, ::TransformStreamDefaultController::stream(controller));
+
+    // 2.  Let readableController be stream.[readable].[controller].
+    RootedObject readable(cx, TransformStream::readable(stream));
+    RootedObject readableController(cx, JS::ReadableStreamGetController(cx, readable));
+    MOZ_ASSERT(readableController);
+
+    // 3.  Perform ! [ReadableStreamDefaultControllerClose](readableController).
+    // Note: in https://github.com/whatwg/streams/pull/1029, the spec was changed to
+    // make ReadableStreamDefaultControllerClose (and -Enqueue) return early if
+    // ReadableStreamDefaultControllerCanCloseOrEnqueue is false. SpiderMonkey hasn't been
+    // updated accordingly, so it'll throw an exception instead.
+    // To avoid that, we do the check explicitly. While that also throws an exception, we can
+    // just clear it and move on.
```

Comment:
Comment here on what happens when SpiderMonkey is updated? I guess then we just have a no-op `ClearPendingException` (which is harmless)?

Reply:
I added the following comment:
```c
    // Note that this is future-proof: if SpiderMonkey is updated accordingly, it'll simply
    // stop throwing an exception, and the `else` branch will never be taken.
```